addons/website/models/website.py

```diff
@@ -1340,3 +1341,235 @@ def _get_cached_values(self):
+        def trigram_enumerate_words():
+            """
+            Browses through all words that need to be compared to the search term.
+            It extracts all words of every field associated to models in the fields_per_model parameter.
+            The search is restricted to a records having the non-zero pg_trgm.word_similarity() score.
+            """
+            match_pattern = '\\w{%s,}' % min(4, len(search) - 3)
+            search_term = "'%s'" % escape_psql(search)
+            self.env.cr.execute("SET pg_trgm.word_similarity_threshold TO 0.3")
+            for search_detail in search_details:
+                model_name, fields = search_detail['model'], search_detail['search_fields']
+                model = self.env[model_name]
+                if search_detail.get('requires_sudo'):
+                    model = model.sudo()
+                fields = ['arch_db' if field == 'arch' else field for field in fields if field in model]
+                if model_name == 'website.page':
+                    fields.remove('name') # TODO handle ?
+
+                if self.env.registry.has_unaccent:
+                    unaccent = get_unaccent_wrapper(self.env.cr)
+                    similarities = ["word_similarity(%s, %s)" % (unaccent("%(search)s"), unaccent(field)) for field in fields]
+                    similar = ["%s <%%%% %s" % (unaccent("%(search)s"), unaccent(field)) for field in fields]
+                else:
+                    similarities = ["word_similarity(%s, %s)" % ("%(search)s", field) for field in fields]
+                    similar = ["%s <%%%% %s" % ("%(search)s", field) for field in fields]
+                best_similarity_str = 'GREATEST(%s)' % ', '.join(similarities)
+
+                query = """
+                    SELECT id, {best_similarity_str} AS _best_similarity
+                    FROM {table}
+                    WHERE {where}
+                    ORDER BY _best_similarity desc
+                    LIMIT 1000
+                """.format(
+                    best_similarity_str=best_similarity_str,
+                    table=model._table,
+                    where=' OR '.join(similar),
+                )
+                self.env.cr.execute(query, {'search': search})
+                ids = set([row[0] for row in self.env.cr.fetchall()])
+                if self.env.lang:
+                    if self.env.registry.has_unaccent:
+                        unaccent = get_unaccent_wrapper(self.env.cr)
+                        similarity_str = "word_similarity(%s, %s)" % (unaccent(search_term), unaccent('value'))
+                        similar = "%s <%%%% %s" % (unaccent(search_term), unaccent('value'))
+                    else:
+                        similarity_str = "word_similarity(%s, %s)" % (search_term, 'value')
+                        similar = "%s <%%%% %s" % (search_term, 'value')
+                    names = ['%s,%s' % (escape_psql(model._name), escape_psql(field)) for field in fields]
```

Comment:
This looks like an example of a confusing and incorrect use of `escape_psql`.
`names` is passed as parameters to `cr.execute()` below, and in the query it's used with `=ANY(names)`. This means LIKE pattern characters don't have special meaning here, and escaping them is actually an error that will cause the equality test to fail.

```sql
14=# SELECT id FROM ir_model WHERE model = ANY(ARRAY['ir.config_parameter']);
 id 
----
 35
(1 row)
14=# SELECT id FROM ir_model WHERE model = ANY(ARRAY['ir.config\_parameter']);
 id 
----
(0 rows)
```